### DISCUSSION TITLE
Strict partial loading and branded types without the ability to use similar types

### DISCUSSION THREAD

**Discussion by mrmlnc** [mikro-orm/mikro-orm#5374] [General]
Hello,

In the new version, we have an excellent opportunity (strict partial loading) to protect ourselves from errors due to not selected fields. This functionality is made using branded types.

https://github.com/mikro-orm/mikro-orm/blob/641887245abbded5df2d984fc24c6fe3cea80c9b/packages/core/src/typings.ts#L1083

As I understand it, this is to guard against the case where an entity was loaded with one set of fields (smaller) and then used in a location with a larger number of loaded fields.

Unfortunately, when migrating to the new version, we ran into dozens of places in our large project where a method load a larger set of entity fields, and then needs to pass that entity into a method that only requires a fraction of those fields.

```ts
// project.entity.ts
import { Entity, PrimaryKey, Property } from "@mikro-orm/core";

@Entity()
export class ProjectEntity {
    @PrimaryKey({ autoincrement: true, hidden: true })
    sid: number;

    @Property()
    id: string;

    @Property()
    title: string;
}
```

```ts
// Loaded<ProjectEntity, never, "title" | "sid", never>
const project = await orm.em
  .getRepository(ProjectEntity)
  .findOneOrFail({ id: 'id' }, { fields: ['sid', 'title'] });

function somethingBasedOnProject(project: Loaded<ProjectEntity, never, 'sid'>): void {
    console.log(project.sid);
}

// @ts-expect-error Here project has the sid and title field, but an error occurs due to a mismatch of the branded type.
// Argument of type 'Loaded<ProjectEntity, never, "title" | "sid", never>' is not assignable to parameter of type 'Loaded<ProjectEntity, never, "sid">'.
//   Type 'Loaded<ProjectEntity, never, "title" | "sid", never>' is not assignable to type '{ [__selectedType]?: [ProjectEntity, never, "sid"] | undefined; }'
somethingBasedOnProject(project);
```

Of course, we could use a DTO or just an interface for the required fields, but that only works in simple cases. This will not work in cases where you need to be able to mutate an entity in a method, or the number of fields is large and it is more advantageous to use the familiar `Loaded` type with a simple union-type.

Perhaps you have any recommendations on how to organize such code, or thoughts on how the Loaded type could be improved or provide some helper that allows you to safely require some of the loaded fields.

Thanks.

**Comment by B4nan** [comment]
Hmm, that's a bug for sure, it's not supposed to work like that. I don't think it's about branded types, it's about the tuple it holds.

**Comment by B4nan** [comment]
9de4965dc1ec952b3addc2d902ee5587c2f38f05 should help

**Comment by mrmlnc** [comment]
Thanks! I checked the latest version on our project. It works correctly for all 100+ cases.
